rmw_connext_dynamic_cpp/src/functions.cpp

```diff
@@ -320,111 +454,283 @@ rmw_create_publisher(
     rmw_set_error_string("failed to get participant qos");
     return NULL;
   }
+  // Past this point, a failure results in unrolling code in the goto fail block.
+  rmw_publisher_t * publisher = nullptr;
+  DDS_TypeCode * type_code = nullptr;
+  void * buf = nullptr;
+  DDSDynamicDataTypeSupport * ddts = nullptr;
+  DDS_PublisherQos publisher_qos;
+  DDSPublisher * dds_publisher = nullptr;
+  DDSTopic * topic = nullptr;
+  DDSTopicDescription * topic_description = nullptr;
+  DDS_DataWriterQos datawriter_qos;
+  DDSDataWriter * topic_writer = nullptr;
+  DDSDynamicDataWriter * dynamic_writer = nullptr;
+  DDS_DynamicData * dynamic_data = nullptr;
+  CustomPublisherInfo * custom_publisher_info = nullptr;
+  // Start initializing elements.
+  publisher = rmw_publisher_allocate();
+  if (!publisher) {
+    rmw_set_error_string("failed to allocate memory for publisher");
+    goto fail;
+  }
 
-  DDS_TypeCode * type_code = create_type_code(type_name, members, participant_qos);
+  type_code = create_type_code(type_name, members, participant_qos);
   if (!type_code) {
     // error string was set within the function
-    return NULL;
+    goto fail;
   }
 
-
-  DDSDynamicDataTypeSupport * ddts = new DDSDynamicDataTypeSupport(
-    type_code, DDS_DYNAMIC_DATA_TYPE_PROPERTY_DEFAULT);
+  // Allocate memory for the DDSDynamicDataTypeSupport object.
+  buf = rmw_allocate(sizeof(DDSDynamicDataTypeSupport));
+  if (!buf) {
+    rmw_set_error_string("failed to allocate memory");
+    goto fail;
+  }
+  // Use a placement new to construct the DDSDynamicDataTypeSupport in the preallocated buffer.
+  TRY_PLACEMENT_NEW(
+    ddts, buf,
+    goto fail,
+    DDSDynamicDataTypeSupport, type_code, DDS_DYNAMIC_DATA_TYPE_PROPERTY_DEFAULT);
+  buf = nullptr;  // Only free the casted pointer; don't need the buf pointer anymore.
   status = ddts->register_type(participant, type_name.c_str());
   if (status != DDS_RETCODE_OK) {
     rmw_set_error_string("failed to register type");
-    return NULL;
+    // Delete ddts to prevent the goto fail block from trying to unregister_type.
+    TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(ddts->, DDSDynamicDataTypeSupport)
```

Comment:
> Answered my own question, looks like you use operator . at later points in the code

^, I didn't want to make two, really long named macros, but we could do that too. Also I could just put the whole destructor statement in there, but this is where I ended up.